numpy/core/src/umath/simd.inc.src

```diff
@@ -1075,6 +1096,386 @@ sse2_@kind@_@TYPE@(@type@ * ip, @type@ * op, const npy_intp n)
 
 /**end repeat**/
 
+/* bunch of helper functions used in ISA_exp/log_FLOAT*/
+
+#if HAVE_ATTRIBUTE_TARGET_AVX2
+NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_AVX2 __m256
+avx2_fmadd(__m256 a, __m256 b, __m256 c)
+{
+    return _mm256_add_ps(_mm256_mul_ps(a, b), c);
```

Comment:
avx2 has fmadd natively, does it not have the required semantics for this and avx512 fmadd does?

Reply:
It does, but it additionally requires "fma" attribute. FMA was introduced post AVX2, so it wouldn't work with the older CPU's which have AVX2 but no FMA.

Reply:
hm right
Though I wonder if this has an effect on the accuracy bounds you quoted for avx2

Reply:
The max ULP error and relative error reported above actually holds true for both AVX2 and AVX512 implementations. They were computed by comparing the output of these functions to glibc's scalar implementation of exp and log in double precision (higher precision) across all 32-bit floats (2^32 values). I did that for both AVX2 and AVX512 versions and turns out the max error is the same for both of them.

Reply:
But the two variants do probably not return the exact same results?
For reproducibility it would be preferable to minimize the amount of different results on different hardware generations.
Only supporting this for AVX2+fma hardware is probably not unreasonable.

Though do you know how common avx2 hardware without fma is?

Reply:
So, I did run an exhaustive check for all float32 and I verified that avx2/avx512 version of log and exp do produce the exact same result (0 ULP error, 0.0 absolute error). 

Intel introduced FMA in Haswell. So, Ivy Bridge and Sandy Bridge (released in 2011-2012) don't have FMA's. Since they seem to be producing the same result, I would like to keep it as is if that is okay with you. 

> very nice, can you provide the code you used to compare the implementations?
> 
> please do not add sin/cos to this PR, lets do that one at the time.